common/membership/hashring_test.go

```diff
@@ -116,27 +116,85 @@ func TestRefreshUpdatesRingOnlyWhenRingHasChanged(t *testing.T) {
 	pp := NewMockPeerProvider(ctrl)
 
 	pp.EXPECT().Subscribe(gomock.Any(), gomock.Any()).Times(1)
-	pp.EXPECT().GetMembers("test-service").Times(3)
+	pp.EXPECT().GetMembers("test-service").Times(1).Return(randomHostInfo(3), nil)
 
 	hr := newHashring("test-service", pp, log.NewNoop(), metrics.NoopScope(0))
+	// Start will also call .refresh()
 	hr.Start()
-
-	hr.refresh()
 	updatedAt := hr.members.refreshed
 	hr.refresh()
+	refreshed, err := hr.refresh()
+
+	assert.NoError(t, err)
+	assert.False(t, refreshed)
 	assert.Equal(t, updatedAt, hr.members.refreshed)
 
 }
 
+func TestRefreshWillNotifySubscribers(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	pp := NewMockPeerProvider(ctrl)
+	var hostsToReturn []HostInfo
+	hostLock := &sync.RWMutex{}
+	pp.EXPECT().Subscribe(gomock.Any(), gomock.Any()).Times(1)
+	pp.EXPECT().GetMembers("test-service").Times(2).DoAndReturn(func(service string) ([]HostInfo, error) {
+		hostLock.Lock()
+		hostsToReturn = randomHostInfo(5)
+		hostLock.Unlock()
+		time.Sleep(time.Millisecond * 70)
+		return hostsToReturn, nil
+	})
+
+	changed := &ChangedEvent{
+		HostsAdded:   []string{"a"},
+		HostsUpdated: []string{"b"},
+		HostsRemoved: []string{"c"},
+	}
+
+	hr := newHashring("test-service", pp, log.NewNoop(), metrics.NoopScope(0))
+	hr.Start()
+
+	var changeCh1 = make(chan *ChangedEvent)
+	var changeCh2 = make(chan *ChangedEvent)
+
+	assert.NoError(t, hr.Subscribe("notifyMe1", changeCh1))
+	assert.NoError(t, hr.Subscribe("notifyMe2", changeCh2))
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		changedEvent := <-changeCh1
+		close(changeCh1)
+		assert.Equal(t, changed, changedEvent)
+	}()
```

Comment:
nothing guarantees these goroutines reach the `<-changeCh1` read before the the `hr.refreshChan <- changed` occurs, which would lead to missing the notification.

probably just buffer these chans?